Add context object

brentzundel opened the pull request:
Adds ledger transactions to add and retrieve a context object as documented in [The rich schema hipe](https://github.com/hyperledger/indy-hipe/tree/master/text/0119-rich-schemas) and the [rich schema context hipe](https://github.com/hyperledger/indy-hipe/tree/master/text/0138-rich-schema-context)

ken-ebert commented:
@ashcherbakov The build errors seem to be unrelated to the code changed by this PR. 
"Total 1 runs 2603 passed, 1 failed, 0 errors, 38 skipped
Failed tests:
indy_node/test/anon_creds:test_state_proof_returned_for_delta_with_None_reply
Writing failure data in Test-Report.txt
Tests run. Returning 256"

ashcherbakov commented:
@ken-ebert 
The test must pass on Sovrin Jenkins only, so don't worry, all the tests passed for you PR.
There are some intermittent tests that fail on Hyperledger Jenkins. This is a known issue and we have a TODO task to fix this.

BTW I've started review of the PR (looks good so far), planning to continue and put a feedback tomorrow.

ken-ebert commented:
Thanks for the update. And thanks for your review. We will look closely at
your suggestions as you are complete your review.
Ken

On Tue, Sep 3, 2019 at 10:59 AM ashcherbakov <notifications@github.com>
wrote:

> @ken-ebert <https://github.com/ken-ebert>
> The test must pass on Sovrin Jenkins only, so don't worry, all the tests
> passed for you PR.
> There are some intermittent tests that fail on Hyperledger Jenkins. This
> is a known issue and we have a TODO task to fix this.
>
> BTW I've started review of the PR (looks good so far), planning to
> continue and put a feedback tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperledger/indy-node/pull/1430?email_source=notifications&email_token=AKFZ3CWAVKNOPKZVDT2OGBLQH2JYXA5CNFSM4IRTVZI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5Y3X6I#issuecomment-527547385>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKFZ3CXFBO4VO54IXCD3HEDQH2JYXANCNFSM4IRTVZIQ>
> .
>

ashcherbakov commented on indy_node/server/request_handlers/domain_req_handlers/context_handler.py:
```diff
@@ -0,0 +1,134 @@
+    return False
+
+
+def validate_data(data):
```
`validate_data` and `validate_meta` looks like Schema-based static validation, so it's better to do it there (within `ClientSetContextOperation` in `types.py`).
In general the static validation should be done as follows:
- validation that we have expected fields of expected types should be done as Schema-based validation in `types.py`
- request-specific business logic which doesn't depend on the current state and order of transactions should be done in `static_validation` in Request Handler classes.

brentzundel replied:
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

ashcherbakov commented on indy_node/server/request_handlers/domain_req_handlers/context_handler.py:
```diff
@@ -0,0 +1,134 @@
+
+def validate_data(data):
+    if isinstance(data, dict):
+        if "@context" not in data.keys():
```
Please use constants  (`CONTEXT_CONTEXT` for example) instead of strings.

brentzundel replied:
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

ashcherbakov commented on indy_node/server/request_handlers/domain_req_handlers/context_handler.py:
```diff
@@ -0,0 +1,134 @@
+def validate_data(data):
+    if isinstance(data, dict):
+        if "@context" not in data.keys():
+            raise Exception("data missing '@context' property")
```
Please raise more specific exceptions (`InvalidClientRequest` for example)

brentzundel replied:
fixed in [3190909](https://github.com/hyperledger/indy-node/pull/1430/commits/31909099fb629c02c2ec62d4ade5d22ea723b8d3)

ashcherbakov commented on indy_node/server/request_handlers/domain_req_handlers/context_handler.py:
```diff
@@ -0,0 +1,134 @@
+
+    def dynamic_validation(self, request: Request):
+        # we can not add a Context with already existent NAME and VERSION
+        # since a Context needs to be identified by seqNo
```
It would be great to create a Design doc / HIPE describing how reach schema objects are related to each other. 
Does it use `seqNo` or other IDs for referencing? 

Currently Schema is referenced by ClaimDef via `seqNo` on Indy-Node level, but ClaimDef is referenced by RevocRegDef via generic ID mechanism which is currently equal to ClaimDef's key in the state trie, that is unique ClaimDef identifier allowing to potentially update referenced ClaimDef (state can be considered as a dict accumulates all the changes for an object).
On libindy level everything is supposed to be referenced via IDs (including referencing a Schema from a ClaimDef).

ashcherbakov commented on indy_node/test/request_handlers/test_context_handler.py:
```diff
@@ -0,0 +1,267 @@
+
+def test_validate_meta_fail_on_empty():
+    with pytest.raises(KeyError) as e:
+        validate_meta({})
```
I think it would be better to call `ch.static_validation` in this and the following tests to make sure that static validation really throws expected errors. We are testing ContextHandler's API here, and `static_validation` is one of the API methods.

brentzundel replied:
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

ashcherbakov commented on indy_node/test/request_handlers/test_context_handler.py:
```diff
@@ -0,0 +1,267 @@
+    }
+    req = Request("test", 1, operation, "sig", )
+    ch = ContextHandler(None, None)
+    with pytest.raises(KeyError) as e:
```
I'm not sure that `KeyError` is expected error here. I believe expected error should be `InvalidClientRequest`

brentzundel replied:
This test has been moved to `indy_node/test/context/test_send_context.py`  and is called `test_validate_operation_fail_no_meta`
It now expects a `TypeError`

ashcherbakov commented on indy_node/test/request_handlers/test_context_handler.py:
```diff
@@ -0,0 +1,267 @@
+    make_context_exist(context_request, context_handler)
+    with pytest.raises(UnauthorizedClientRequest, match=str(AuthConstraintForbidden())):
+        context_handler.dynamic_validation(context_request)
+
```
Please uncomment/write more unit tests here for `dynamic_validation` and `update_state`

brentzundel replied:
fixed in [3fd8694](https://github.com/hyperledger/indy-node/pull/1430/commits/3fd869442e5b1b4377c8cc295c7707dbacfe66c1)

ashcherbakov commented on indy_node/test/request_handlers/test_context_handler.py:
```diff
@@ -0,0 +1,267 @@
+    with pytest.raises(UnauthorizedClientRequest, match=str(AuthConstraintForbidden())):
+        context_handler.dynamic_validation(context_request)
+
+'''
```
Please also add unit tests for `GetContextReqHandler`

brentzundel replied:
I'm having trouble finding equivalent unit tests for any of the other read request handlers (beyond the tests that call `get_result` in the `anon_creds` and `state_proof` tests, which have been implemented for CONTEXT). 
Could you point me to some?

ashcherbakov replied:
I believe the equivalent unit tests are the ones commented out below:
1) test that dynamic validation passes
2) test that dynamic validation fails
3) test for correct state update

ashcherbakov commented on indy_node/test/context/test_send_get_context.py:
```diff
@@ -0,0 +1,310 @@
+TEST_CONTEXT_NAME = "Base_Context"
+TEST_CONTEXT_VERSION = "1.0"
+
+
```
Please add tests to `indy_node/test/state_proof/test_state_proof_for_get_requests.py` and `indy_node/test/state_proof/test_state_proof_for_missing_data.py` checking that we can query existing (and non-existing) Contexts from 1 Node only (using BLS multi-sig and state proofs).
We may have to skip the tests until Context is supported in Indy SDK.

brentzundel replied:
I added a test to `indy_node/test/state_proof/test_state_proof_for_get_requests.py` for `GET_CONTEXT` in [033c7e3](https://github.com/hyperledger/indy-node/pull/1430/commits/033c7e31cdd8f6592c4115c5d89d1763fe80dbb5)
As you will see, it uses `nodeSet` rather than `nodeSetWithOneNodeResponding`
It seems to pass all of the assertions, but if I use `nodeSetWithOneNodeResponding` I get a `PoolLedgerTimeout`
Any thoughts as to what the problem might be?

ashcherbakov replied:
As I mentioned above, this will not work until Context object is supported on SDK side (that is SDK has a logic for verification of BLS multi-signed state proofs for the Context object). Since SDK doesn't have Context object support yet, it always fallbacks to read from all Nodes, and reading from 1 Node raises `PoolLedgerTimeout`.

ashcherbakov commented on indy_common/authorize/auth_map.py:
```diff
@@ -68,6 +68,16 @@
                              old_value='*',
                              new_value='*')
 
+add_context = AuthActionAdd(txn_type=SET_CONTEXT,
```
Please extend `https://github.com/hyperledger/indy-node/blob/master/docs/source/auth_rules.md` for new transactions

brentzundel replied:
added in [655aed4](https://github.com/hyperledger/indy-node/pull/1430/commits/655aed4f4f224b429304d343b0c9c69737635bf8)

ashcherbakov commented on indy_common/constants.py:
```diff
@@ -160,6 +171,10 @@
 CHANGE_KEY = IndyTransactions.CHANGE_KEY.value
 
+# Rich Schema
+SET_CONTEXT = IndyTransactions.SET_CONTEXT.value
```
Please extend `https://github.com/hyperledger/indy-node/blob/master/docs/source/transactions.md` and `https://github.com/hyperledger/indy-node/blob/master/docs/source/requests.md` with a description of new requests (SET_CONTEXT and GET_CONTEXT) and transaction (SET_CONTEXT)

brentzundel replied:
added in [20ea2b2](https://github.com/hyperledger/indy-node/pull/1430/commits/20ea2b297ba8969fb8e9cc183e678ca2dbf5411c)

ashcherbakov commented on indy_common/constants.py:
```diff
@@ -160,6 +171,10 @@
 CHANGE_KEY = IndyTransactions.CHANGE_KEY.value
 
+# Rich Schema
+SET_CONTEXT = IndyTransactions.SET_CONTEXT.value
```
Out general pattern for write requests (transaction) is a Noun: NYM, SCHEMA, CLAIM_DEF. etc.
So, I think we may rename `SET_CONTEXT` to `CONTEXT` for consistency.
Also I think that `CONTEXT` may sound too generic. Maybe we can call it `SCHEMA_CONTEXT`?

brentzundel replied:
We can't call it `SCHEMA_CONTEXT` because context objects may be used by more than just schemas.
As far as changing it from `SET_CONTEXT` to simply `CONTEXT` I believe @ken-ebert had good reason for `SET_CONTEXT` and perhaps could add some comments here.

ashcherbakov commented on indy_node/test/context/helper.py:
```diff
@@ -0,0 +1,5393 @@
+
+
```
Do we have integration tests trying to write these Context examples?

brentzundel replied:
These examples are used in `test_send_context.py` and `test_send_get_context.py`

ashcherbakov commented:
@ken-ebert @brentzundel 
Thank you for your contribution! In general it looks good, and there are no any critical things. 
Please find a summary of findings (look at the comments in PR for details):
1) I think it's better to rename `SET_CONTEXT` to just `CONTEXT` (or better `SCHEMA_CONTEXT`)
2) The format of SET_CONTEXT and GET_CONTEXT differs from the Context HIPE. Also there are some unused fields there
3) Increase a general limit for message size in Plenum if Context object is really close to it.
4) Think about moving general static validation checks to schema-based validation in `types.py`
5) `None` needs to be returned in case of absent data for GET_CONTEXT
6) Add more unit tests for Set Context and Get Context request handlers
7) Add tests for getting context objects from 1 Node only (we may have to skip them until it's supported on SDK side)
8) Integration tests for W3C context examples
9) Extend documentation 
10) Think about relationship of Rich Schema objects: reference by seqNo (maybe not so flexible), or by ID? How ID may look like (state trie key based ID as of now, or a different format)

ken-ebert commented on indy_common/config.py:
```diff
@@ -103,3 +103,5 @@
 
 SCHEMA_ATTRIBUTES_LIMIT = 125
+
+CONTEXT_SIZE_LIMIT = 131072
```
We have a test that exceeds our limit for a context CONTEXT_SIZE_LIMIT, but does not seem to trigger the MSG_LEN_LIMIT during the Plenum check. Instead we get a PoolLedgerTimeoutException.  Is there compression of messages that would reduce the size of the message at the Plenum level?

The largest context we have seen so far is from schema.org at 121278 bytes. What do you recommend?

ashcherbakov replied:
>  Instead we get a PoolLedgerTimeoutException
Yes, I believe this is what expected from the client side if message is discarded because of `MSG_LEN_LIMIT ` checks.
`MSG_LEN_LIMIT ` checks are done on zstack level (network level) before parsing and handling request, so for the client it looks like a timeout.
Please note that `MSG_LEN_LIMIT ` checks are done before decoding the message (have a look at `_verifyAndAppend` in plenum's `zstack.py`).

I recommend just to increase `MSG_LEN_LIMIT ` in Plenum.

brentzundel commented on indy_common/state/domain.py:
```diff
@@ -4,7 +4,7 @@
 from indy_common.constants import ATTRIB, GET_ATTR, SIGNATURE_TYPE, REVOC_TYPE, TAG, CRED_DEF_ID, REVOC_REG_DEF_ID, \
     CLAIM_DEF_SCHEMA_REF, CLAIM_DEF_PUBLIC_KEYS, SCHEMA_ATTR_NAMES, CLAIM_DEF_FROM, CLAIM_DEF_TAG, \
-    CLAIM_DEF_TAG_DEFAULT, CLAIM_DEF_CL
+    CLAIM_DEF_TAG_DEFAULT, CLAIM_DEF_CL, CONTEXT_CONTEXT
```
fixed in [c4a4242](https://github.com/hyperledger/indy-node/pull/1430/commits/c4a4242a0257b2e6581ba4579e3ea30e249c0f67)

ken-ebert commented on indy_common/types.py:
```diff
@@ -93,6 +99,58 @@ class SchemaField(MessageValidator):
+        self.max_size = max_size
+        super().__init__(**kwargs)
+
+    def _specific_validation(self, val):
```
I think that the validation for context size should be independent of and in addition to the overall message maximum size check conducted by Plenum. The context node code should be able to validate that a context is within the size restrictions for a context object.

ashcherbakov replied:
If we can predict such a maximum size for a Context object - then fine.

brentzundel commented on indy_node/test/context/test_send_get_context.py:
```diff
@@ -0,0 +1,310 @@
+def test_send_get_context_succeeds(
+        looper, sdk_pool_handle, nodeSet, sdk_wallet_trustee, send_context):
+    _, did = sdk_wallet_trustee
+    raw_json = {
```
It is not clear to me why Indy-SDK should be involved in testing Indy-Node functionality.

ashcherbakov replied:
indy-sdk is not involved in unit tests, but it's involved in integration tests where we send real requests.
The reasons for this:
1) Historical/legacy reasons
2) Greatly improve the confidence in compatibility between Node and SDK
3) No need to have separate client code in Indy-Node
4) This is OK for integration tests (that's one of the reasons why they are integration)

brentzundel replied:
but shouldn't integration tests which involve indy-sdk live in the indy-sdk code?
Indy-sdk requires indy-node, but it is possible to use indy-node without indy-sdk.
How will integration tests that currently rely on indy-sdk be changed once aries becomes functional? Will they be changed to rely on aries? This doesn't make sense to me.

ashcherbakov replied:
Indy-SDK has integration tests as well.
There are things that can be done much easier in Plenum/Node, not in SDK. State proof reads (reads from 1 Node) is one of the examples.
I believe yes, once Indy-SDK is fully deprecated, Plenum and Node needs to be switched to Aries.

brentzundel replied:
It seems completely inappropriate to assume inside of indy-node which library will be used (i.e., indy-SDK, aries, etc.) to connect to it. 
I do not understand this at all. In my opinion, every test in indy-node that relies on indy-SDK should be changed to remove that dependency.

brentzundel commented on indy_common/test/types/test_context_schema.py:
```diff
@@ -0,0 +1,56 @@
+    ClientSetContextOperation, ContextField
+from plenum.common.messages.fields import ConstantField, VersionField, IterableField, LimitedLengthStringField
+
+EXPECTED_ORDERED_META_FIELDS_SCHEMA = OrderedDict([
```
fixed in [abb035d](https://github.com/hyperledger/indy-node/pull/1430/commits/abb035da26451854c224945bd4205e55722c52b7)

brentzundel commented on indy_common/test/types/test_get_context_schema.py:
```diff
@@ -0,0 +1,39 @@
+from indy_common.types import ClientGetContextOperation, GetContextField
+from plenum.common.messages.fields import ConstantField, IdentifierField, VersionField, LimitedLengthStringField
+
+EXPECTED_ORDERED_FIELDS_SCHEMA = OrderedDict([
```
fixed in [abb035d](https://github.com/hyperledger/indy-node/pull/1430/commits/abb035da26451854c224945bd4205e55722c52b7)

brentzundel commented on indy_common/types.py:
```diff
@@ -93,6 +99,58 @@ class SchemaField(MessageValidator):
+class ContextField(FieldBase):
+    _base_types = None
+
+    def __init__(self, inner_field_type: FieldValidator, max_size=None, **kwargs):
```
It is currently not used. [97ddd0d](https://github.com/hyperledger/indy-node/pull/1430/commits/97ddd0d28003566176cf046a8ff27dad15fa68e1) removed it.
Originally the plan was to use it for schema-based static validation of the value of the CONTEXT_CONTEXT property, but I wasn't sure how to use schema-based validation to do so, as the value may be an array, a dictionary, or a URI.
This level of static validation is currently done by the context_handler.

brentzundel commented on indy_common/types.py:
```diff
@@ -93,6 +99,58 @@ class SchemaField(MessageValidator):
+
+class SetContextDataField(MessageValidator):
+    schema = (
+        (CONTEXT_CONTEXT, ContextField(
```
yes, we need to update the HIPE

brentzundel commented on indy_common/types.py:
```diff
@@ -93,6 +99,58 @@ class SchemaField(MessageValidator):
+    )
+
+
+class ContextField(FieldBase):
```
ContextField needs to be either an array, a dictionary, or a URI (including a context identifier). I'm not sure how to check for this using schema-based validation, so for now we only check the size of the schema to make sure it is not too big.

brentzundel replied:
and validate the value of the context in the context_handler static validation.

brentzundel commented on indy_common/types.py:
```diff
@@ -155,6 +213,24 @@ class ClientGetSchemaOperation(MessageValidator):
+class ClientGetContextOperation(MessageValidator):
+    schema = (
+        (TXN_TYPE, ConstantField(GET_CONTEXT)),
+        (CONTEXT_FROM, IdentifierField()),
```
Yes, we need to update the HIPE

brentzundel commented on indy_common/types.py:
```diff
@@ -155,6 +213,24 @@ class ClientGetSchemaOperation(MessageValidator):
+# Rich Schema
+# this class is not actually used for static validation at this time
+class ClientSetContextOperation(MessageValidator):
+    schema = (
```
Yes, we need to update the HIPE

brentzundel commented on indy_common/types.py:
```diff
@@ -155,6 +213,24 @@ class ClientGetSchemaOperation(MessageValidator):
+class ClientGetContextOperation(MessageValidator):
+    schema = (
+        (TXN_TYPE, ConstantField(GET_CONTEXT)),
+        (CONTEXT_FROM, IdentifierField()),
```
What is the difference between `ORIGIN` from `GetSchemaField` and and `SCHEMA_FROM` ?

ashcherbakov replied:
I believe this is just some legacy code and `ORIGIN` in `GetSchemaField` is not used (see https://github.com/hyperledger/indy-node/blob/master/docs/source/requests.md#get_schema)

brentzundel replied:
We copied the context schema-based validation from the schema schema-based validation, so if the schema schema-based validation is not correct (i.e., it checks for `ORIGIN`) then perhaps that needs to be fixed.

ashcherbakov replied:
`ORIGIN` is an optional field in Schema, that's why validation passes there. But in general yes, `ORIGIN` needs to be removed from Schema as well.

brentzundel replied:
Fixed in [40ce80d](https://github.com/hyperledger/indy-node/pull/1430/commits/40ce80d1a525b1a631e8be8ecdcad7dddaaea4ac)

brentzundel commented on indy_node/server/request_handlers/domain_req_handlers/context_handler.py:
```diff
@@ -0,0 +1,134 @@
+
+    def static_validation(self, request: Request):
+        self._validate_request_type(request)
+        meta = request.operation['meta']
```
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

brentzundel commented on indy_node/server/request_handlers/domain_req_handlers/context_handler.py:
```diff
@@ -0,0 +1,134 @@
+    def static_validation(self, request: Request):
+        self._validate_request_type(request)
+        meta = request.operation['meta']
+        validate_meta(meta)
```
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

brentzundel commented on indy_node/test/api/helper.py:
```diff
@@ -216,3 +218,72 @@ def sdk_write_schema_and_check(looper, sdk_pool_handle, sdk_wallet_client,
+    set_context_txn_json = json.dumps(raw_json)
+
+
+    '''request = looper.loop.run_until_complete(
```
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

brentzundel commented on indy_node/test/context/test_send_get_context.py:
```diff
@@ -0,0 +1,310 @@
+    assert rep[0][1]['result']['data']['data'] == W3C_BASE_CONTEXT
+
+
+def test_send_get_context_as_client(
```
it tests that a context written by a trustee may be retrieved by a DID from the client wallet

brentzundel commented on indy_node/test/context/test_send_get_context.py:
```diff
@@ -0,0 +1,310 @@
+    with pytest.raises(RequestNackedException) as e:
+        sdk_get_and_check_replies(looper, [sdk_sign_and_submit_req(sdk_pool_handle, sdk_wallet_trustee,
+                                                                   get_context_txn_json)])
+    e.match("Invalid version: 'asd'")
```
fixed in [51fd316](https://github.com/hyperledger/indy-node/pull/1430/commits/51fd316b9346906f925fe1cd0f08780983f9387f)

ashcherbakov commented on docs/source/auth_rules.md:
```diff
@@ -285,6 +285,24 @@
+    <td><sub><code>*</code></sub></td>
+    <td><sub><code>*</code></sub></td>
+    <td><sub>1 TRUSTEE OR 1 STEWARD OR 1 ENDORSER</sub></td>
+    <td><sub>Adding a new Schema</sub></td>
```
I believe it needs to be `Adding a new Context`

brentzundel replied:
fixed in [ff6f6d7](https://github.com/hyperledger/indy-node/pull/1430/commits/ff6f6d7db1275f4eaadf43cf17127f8c82390936)

ashcherbakov commented on docs/source/auth_rules.md:
```diff
@@ -285,6 +285,24 @@
+    <td><sub><code>*</code></sub></td>
+    <td><sub><code>*</code></sub></td>
+    <td><sub>No one can edit existing Context</sub></td>
+    <td><sub>Editing a Schema</sub></td>
```
I believe it needs to be `Editing a Context`

brentzundel replied:
fixed in [ff6f6d7](https://github.com/hyperledger/indy-node/pull/1430/commits/ff6f6d7db1275f4eaadf43cf17127f8c82390936)

ashcherbakov commented on docs/source/auth_rules.md:
```diff
@@ -548,6 +566,18 @@
+  <tr>
+    <td><sub>SET_CONTEXT</sub></td>
+    <td><sub>EDIT</sub></td>
+    <td><sub>A CONTEXT cannot be edited</sub></td>
```
This spreadsheet clarifies who is the Owner of the object regardless whether it can be edited (and BTW whether an object can be edited is defined by Auth Rules which can be customized by a Network Administrators; so can-not-edit default rules for Schema and Context can be changed (in theory) to the ones allowing editing). 

So, I believe something like `The DID used to create the CONTEXT` needs to be there.

brentzundel replied:
fixed in [ff6f6d7](https://github.com/hyperledger/indy-node/pull/1430/commits/ff6f6d7db1275f4eaadf43cf17127f8c82390936)

ashcherbakov commented on docs/source/requests.md:
````diff
@@ -1882,6 +1888,110 @@ Each acceptance mechanisms list has a unique version.
 }
 ```
 
+### GET_CONTEXT
````
Should it be `SET_CONTEXT`?

ashcherbakov commented on docs/source/requests.md:
```diff
@@ -1882,6 +1888,110 @@ Each acceptance mechanisms list has a unique version.
+ 
+     Dictionary with Context's data:
+     
+    - `attr_names`: array of attribute name strings (125 attributes maximum)
```
These looks like fields for Schema, not for Context.

ashcherbakov commented on indy_node/test/state_proof/test_state_proof_for_get_requests.py:
```diff
@@ -146,6 +150,120 @@ def test_state_proof_returned_for_get_schema(looper,
     check_valid_proof(result)
 
 
+def test_state_proof_returned_for_get_context(looper,
```
Since both `test_state_proof_returned_for_get_context` are in one module, I believe `nodeSet` fixtures is the same as `nodeSetWithOneNodeResponding`, since `nodeSet` is a module-level fixture, and `nodeSetWithOneNodeResponding` extends it (so this is references to the same nodes) to delay requests for all nodes but one.
You may need to move this test into another test module, for example `indy_node/test/state_proof/test_state_multi_proofs_for_get_requests.py`.

ashcherbakov commented:
@brentzundel @ken-ebert 
The PR is almost ready for merge. The following is left:
1) There is one test failing currently (`test_state_proof_returned_for_get_context`). Please have a look at my comment above for a description why it is failing (https://github.com/hyperledger/indy-node/pull/1430#discussion_r322615923).
2) Fix https://github.com/hyperledger/indy-node/pull/1430#discussion_r320731667
3) A couple of typos in docs
4) More unit tests would be appreciated: https://github.com/hyperledger/indy-node/pull/1430/files#r320747858 and https://github.com/hyperledger/indy-node/pull/1430/files#r320747983

brentzundel commented on indy_node/server/request_handlers/read_req_handlers/get_context_handler.py:
```diff
@@ -0,0 +1,48 @@
+            with_proof=True
+        )
+        if context is None:
+            context = {}
```
fixed in [762071b](https://github.com/hyperledger/indy-node/pull/1430/commits/762071b3bc540a1ad4de12a97f0706397d6d860e)

brentzundel commented on indy_node/test/api/helper.py:
```diff
@@ -216,3 +218,72 @@ def sdk_write_schema_and_check(looper, sdk_pool_handle, sdk_wallet_client,
+            'data': context
+        },
+        "identifier": did,
+        "reqId": 12345678,
```
I use the passed in reqId to test both cases.

brentzundel commented on indy_node/test/api/helper.py:
```diff
@@ -216,3 +218,72 @@ def sdk_write_schema_and_check(looper, sdk_pool_handle, sdk_wallet_client,
+        },
+        "identifier": did,
+        "reqId": 12345678,
+        "protocolVersion": 2,
```
fixed in [fce8f58](https://github.com/hyperledger/indy-node/pull/1430/commits/fce8f58469466364b1080f0bdfb3b3a0afc6a722)

brentzundel commented:
@ashcherbakov I believe I have addressed all of the concerns you had during your (excellent and much appreciated) review.
Please let us know if there is anything else keeping this PR from being mergable.

ashcherbakov commented on docs/source/requests.md:
```diff
@@ -2843,6 +2964,96 @@ All input parameters are optional and mutually exclusive.
+    'operation': {
+        'type': '300'
+        'dest': '2VkbBskPNNyWrLrZq7DBhk',
+        'data': {
```
```suggestion
        'meta': {
```

ashcherbakov commented:
@brentzundel 
Thanks for processing all review comments. 
I'm fine with merging the PR, it looks really good. The only thing which prevents it is 1 failing test: `test_state_proof_returned_for_missing_context`. I believe it fails with the same reason as `test_state_proof_returned_for_get_context` used to fail: it requires context support in libindy. So, we can skip `test_state_proof_returned_for_missing_context` for a while.
